pkg/workflows/sdk/v2/runtime.go

```diff
@@ -0,0 +1,95 @@
+package sdk
+
+import (
+	"errors"
+	"io"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/sdk/v2/pb"
+)
+
+type RuntimeBase interface {
+	// CallCapability is meant to be called by generated code
+	CallCapability(request *pb.CapabilityRequest) Promise[*pb.CapabilityResponse]
+	Config() []byte
+	LogWriter() io.Writer
+}
+
+type NodeRuntime interface {
+	RuntimeBase
+	IsNodeRuntime()
+}
+
+type DonRuntime interface {
+	RuntimeBase
+
+	// RunInNodeMode is meant to be used by the helper method RunInNodeMode
+	RunInNodeMode(fn func(nodeRuntime NodeRuntime) *pb.BuiltInConsensusRequest) Promise[values.Value]
+}
+
+type PrimitiveConsensusWithDefault[T any] struct {
+	pb.SimpleConsensusType
+	DefaultValue T
+}
+
+type BuiltInConsensus[T any] interface {
+	pb.SimpleConsensusType | *PrimitiveConsensusWithDefault[T]
+}
+
+var nodeModeCallInDonMode = errors.New("cannot use NodeRuntime outside RunInNodeMode")
+
+func NodeModeCallInDonMode() error {
+	return nodeModeCallInDonMode
+}
+
+var donModeCallInNodeMode = errors.New("cannot use the DonRuntime inside RunInNodeMode")
+
+func DonModeCallInNodeMode() error {
+	return donModeCallInNodeMode
+}
+
+func RunInNodeMode[T any, C BuiltInConsensus[T]](runtime DonRuntime, fn func(nodeRuntime NodeRuntime) (T, error), consensus C) Promise[T] {
+	observationFn := func(nodeRuntime NodeRuntime) *pb.BuiltInConsensusRequest {
+		var primitiveConsensus *pb.PrimitiveConsensus
+		var defaultValue values.Value
+		var err error
+		switch c := any(consensus).(type) {
```

Comment:
is it necessary to cast to `any` here?

Reply:
Yes, otherwise there's a compiler error:

`Cannot use type switch on type parameter value consensus`